File: src/main/scala/com/github/fherbreteau/gatling/sftp/SftpDsl.scala
package com.github.fherbreteau.gatling.sftp

import com.github.fherbreteau.gatling.sftp.protocol.{SftpProtocol, SftpProtocolBuilder}
import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression

import scala.language.implicitConversions

trait SftpDsl {

  def sftp(implicit configuration: GatlingConfiguration): SftpProtocolBuilder = SftpProtocolBuilder(configuration)

  def sftp(requestName: Expression[String]): Sftp = Sftp(requestName)

  implicit def sftpProtocolBuilder2SftpProtocol(builder: SftpProtocolBuilder): SftpProtocol = builder.build
}

File: src/main/java/com/github/fherbreteau/gatling/sftp/javaapi/SftpDsl.java
package com.github.fherbreteau.gatling.sftp.javaapi;

import com.github.fherbreteau.gatling.sftp.javaapi.protocol.SftpProtocolBuilder;
import io.gatling.javaapi.core.Session;
import io.gatling.javaapi.core.internal.Expressions;

import javax.annotation.Nonnull;
import java.util.function.Function;

public class SftpDsl {

    public static final SftpProtocolBuilder sftp =
            new SftpProtocolBuilder(com.github.fherbreteau.gatling.sftp.protocol.SftpProtocolBuilder.apply(
                    io.gatling.core.Predef.configuration()));

    private SftpDsl() {
    }

    @Nonnull
    public static Sftp sftp(@Nonnull String name) {
        return new Sftp(Expressions.toStringExpression(name));
    }

    @Nonnull
    public static Sftp sftp(@Nonnull Function<Session, String> name) {
        return new Sftp(Expressions.javaFunctionToExpression(name));
    }

}
